src/java.base/share/classes/java/lang/String.java

```diff
@@ -553,135 +553,141 @@ public String(byte[] bytes, int offset, int length, Charset charset) {
      * disambiguate it against other similar methods of this class.
      */
     private String(Charset charset, byte[] bytes, int offset, int length) {
+        String str;
         if (length == 0) {
-            this.value = "".value;
-            this.coder = "".coder;
+            str = "";
         } else if (charset == UTF_8.INSTANCE) {
-            if (COMPACT_STRINGS) {
-                int dp = StringCoding.countPositives(bytes, offset, length);
-                if (dp == length) {
-                    this.value = Arrays.copyOfRange(bytes, offset, offset + length);
-                    this.coder = LATIN1;
-                    return;
+            str = utf8(bytes, offset, length);
+        } else if (charset == ISO_8859_1.INSTANCE) {
+            str = iso88591(bytes, offset, length);
+        } else if (charset == US_ASCII.INSTANCE) {
+            str = ascii(bytes, offset, length);
+        } else {
+            str = create(charset, bytes, offset, length);
+        }
+        this.value = str.value;
+        this.coder = str.coder;
```

Comment:
```suggestion
        this(str);
```
We have integrated flexible constructor bodies.

Reply:
```
    public String(String original) {
        this.value = original.value;
        this.coder = original.coder;
        this.hash = original.hash;
        this.hashIsZero = original.hashIsZero;
    }
```
this(str) It will lead to more operations, which are not necessary here.

Reply:
You have shown C2 eliminates this string allocation. Maybe it eliminates these "more operations" too? Imo not worth maintaining separate logic here; this will be problematic when string fields change in the future and we may miss this site that need to update. The simple copy constructor has no issue with this.